---
output:
  pdf_document: default
  html_document: default
---
# Perform Differential Expression analysis using **DESeq2** 

# Loading Required Libraries
```{r message=FALSE}
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(tinytex)
library(magrittr)
```

# Read expression matrix into CSV file.
```{r}
read_count_data <- function(file_path){
counts_data <- read.csv(file_path, row.names = 1)
return(counts_data)
}
expression_matrix <- read_count_data("../DESeq2/Data/pasilla_gene_exp.csv")
```

# Read metadata into CSV file.
```{r}
read_metadata <- function(file_path){
coldata <- read.csv(file_path, row.names = 1)
return(coldata)
}
meta_data <- read_metadata("../DESeq2/Data/pasilla_meta.data.csv")
```

## convert condition and types columns in meta_data object to factor
```{r}
convert_chr_to_factor <- function(){
meta_data$condition <- factor(meta_data$condition)
meta_data$type <- factor(meta_data$type)
}
convert_chr_to_factor()
```


## make sure the row names in meta_data matches to the column names in expression matrix
```{r}
all(rownames(meta_data) %in% colnames(expression_matrix))
```

## IS the columns of the expression matrix and the rows of the meta_data (information about samples) are in the same order?

```{r}
all(rownames(meta_data) == colnames(expression_matrix))
```

## if Not, make them in the same order.

```{r eval=FALSE}
expression_matrix <- expression_matrix[, rownames(meta_data)]
all(rownames(meta_data) == colnames(expression_matrix))
```

# Pre-filtering: removing rows with low gene counts
## keep rows that have at least 10 reads total
```{r}
pre_filter <- function(){
  # Only keep rows that have total counts above the cutoff
  keep <- expression_matrix %>% rowSums(.) >= 10
  filtered_counts <- expression_matrix[keep,]
  return(filtered_counts)
}
filtered_expression_counts <- pre_filter()
head(filtered_expression_counts,2)
```

# Construct a DESeqDataSet.

```{r}
dds <- function(){
deseqdataset <- DESeqDataSetFromMatrix(countData = filtered_expression_counts, 
                                       colData = meta_data, 
                                       design = ~ condition)
return(deseqdataset)
}
deseqdataset <- dds()
deseqdataset
```

# Differential expression analysis

```{r}
diff_expr_analysis <- function(){
deseqdataset <- DESeq(deseqdataset)
result <- results(deseqdataset, alpha = 0.01 , lfcThreshold = 1.5)
return(result)
}
deseq_result <- diff_expr_analysis()
deseq_result
```

# Top 10 differentail expressed genes
```{r}
ordered_result <- deseq_result[order(deseq_result$padj, decreasing = FALSE), ]
top10 <- head(ordered_result, n=10)
top10
```

# Explore results
```{r}
summary(deseq_result)
```

# Write results to CSV file
```{r}
write_sig_genes <- function(out_path){
write.csv(ordered_result, file = out_path)
}
write_sig_genes("../DESeq2//Output/Significant genes.csv")
```


# Visualizing the results

## PCA plot
```{r}
pca_plot <- function(){
normalized = normTransform(deseqdataset)
jpeg("../DESeq2/Output/PCA.jpeg")
p <- plotPCA(normalized, intgroup=c("condition","type"))
print(p)
dev.off()
}
pca_plot()
```

## MA-plot
```{r}
ma_plot <- function(){
jpeg("../DESeq2/Output/MAplot.jpeg")
plotMA(deseq_result)
dev.off()
}  
ma_plot()
```

## Plot counts

Here we specify the gene which had the smallest padj from the results table
```{r}
plot_counts <- function(){
jpeg("../DESeq2/Output/plot_count.jpeg")
plotCounts(deseqdataset, gene = which.min(deseq_result$padj), intgroup = "condition")
dev.off()
}
plot_counts()
```

## Heatmap
```{r}
heatmap <- function(){
select <- order(rowMeans(counts(deseqdataset)), 
                decreasing = FALSE)[1:20]
df <- as.data.frame(colData(deseqdataset)[,c("condition","type")])
jpeg("../DESeq2/Output/Heatmap.jpeg")
p <- pheatmap::pheatmap(assay(deseqdataset)[select,], cluster_rows = FALSE,
                   show_rownames = FALSE, cluster_cols = FALSE, 
                   annotation_col = df)
print(p)
dev.off()
}
heatmap()
```

## Volcano plot
```{r}
vplcano_plot <- function(){
result.df <- as.data.frame(deseq_result)
result.df$diffexpressed <- "NO"
result.df$diffexpressed[result.df$log2FoldChange > 1.5 & 
                        result.df$padj < 0.01] <- "UP"
result.df$diffexpressed[result.df$log2FoldChange < -1.5 & 
                        result.df$padj < 0.01] <- "DOWN"
jpeg("../DESeq2/Output/Volcano.jpeg")
g <- ggplot(data = result.df, aes(x = log2FoldChange, 
                                  y = -log10(pvalue), 
                                  col = diffexpressed))+ 
  geom_point()+ 
  theme_minimal()+
  geom_vline(xintercept = c(-1.5, 1.5), col = "black", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.01), col = "black", linetype = 'dashed') + 
  scale_color_manual(values = c("#00AFBB", "grey", "#FFDB6D"), 
  labels = c("Downregulated", "Not significant", "Upregulated"))
print(g)
dev.off()
}
vplcano_plot()
```











