---
output:
  pdf_document: default
  html_document: default
---
# Perform Differential Expression analysis using **DESeq2** package

# Load Libraries
```{r message=FALSE}
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(tinytex)

```

# Read counts data

```{r}
read_count_data <- function(file_path){
counts_data <<- read.csv(file_path, row.names = 1)
return (head(counts_data))
}

read_count_data("pasilla_gene_exp.csv")

```
# Read Sample metadata

```{r}
read_metadata <- function(file_path){
coldata <<- read.csv(file_path, row.names = 1)
return (coldata)
}

read_metadata("pasilla_meta.data.csv")
```

## convert condition and types columns in coldata object to factor

```{r}
convert_chr_to_factor <- function(){
coldata$condition <- factor(coldata$condition)
coldata$type <- factor(coldata$type)
}

convert_chr_to_factor()
```


## make sure the row names in colData matches to the column names in counts_data

```{r}
all(rownames(coldata) %in% colnames(counts_data))

```

## IS the columns of the count matrix and the rows of the colData (information about samples) are in the same order?

```{r}
all(rownames(coldata) == colnames(counts_data))
```

## if Not, make them in the same order.

```{r eval=FALSE}

counts_data <- counts_data[, rownames(coldata)]
all(rownames(coldata) == colnames(counts_data))

```


# Construct a DESeqDataSet.

```{r}
deseqdataset <- DESeqDataSetFromMatrix(countData = counts_data, 
                                       colData = coldata, 
                                       design = ~ condition)

deseqdataset

```

# Pre-filtering: removing rows with low gene counts
## keep rows that have at least 10 reads total

```{r}
pre_filter <- function(){
keep <- rowSums(counts(deseqdataset)) >=10
deseqdataset <- deseqdataset[keep,]

return (deseqdataset)

}

pre_filter()

```

# Set the factor level

```{r}
factor_level <- function(){

deseqdataset$condition <- relevel(deseqdataset$condition, ref = "untreated")

}

factor_level()

```


# Differential expression analysis

```{r}
diff_expr_analysis <- function(){
deseqdataset <- DESeq(deseqdataset)
result <<- results(deseqdataset)
result01 <<- results(deseqdataset, alpha = 0.01 , lfcThreshold = 1.5)
return (result01)
}

diff_expr_analysis()
```

# Top 10 differentail expression genes

```{r}
ordered_result <- result01[order(result01$padj, decreasing = FALSE), ]
top10 <- head(ordered_result, n=10)
top10

```

# Explore results

```{r}
summary(result)
```


```{r}
summary(result01)
```

## How many adjusted p-values less than 0.01?

```{r}

sum(result01$padj < 0.01 , na.rm = TRUE)

```


# Write results to CSV file
```{r}
write_sig_genes <- function(out_path){
write.csv(ordered_result, file = out_path)
}

write_sig_genes("Significant genes.csv")

```


# Visualizing the results

## MA-plot
```{r}
ma_plot <- function(){
plotMA(result01)
}  

ma_plot()
```

## Plot counts

Here we specify the gene which had the smallest p value from the results table

```{r}
plot_counts <- function(){
plotCounts(deseqdataset, gene = which.min(result01$padj), intgroup = "condition")
}

plot_counts()
```

## Heatmap

```{r}

heatmap <- function(){
select <- order(rowMeans(counts(deseqdataset)), 
                decreasing = TRUE)[1:20]

df <- as.data.frame(colData(deseqdataset)[,c("condition","type")])
pheatmap::pheatmap(assay(deseqdataset)[select,], cluster_rows = FALSE,
                   show_rownames = FALSE, cluster_cols = FALSE, 
                   annotation_col = df)
}

heatmap()

```


## PCA plot

```{r}
pca_plot <- function(){
normalized = normTransform(deseqdataset)
plotPCA(normalized, intgroup=c("condition","type"))
}

pca_plot()
```


## Volcano plot

```{r}
vplcano_plot <- function(){
  
result.df <- as.data.frame(result)

result.df$diffexpressed <- "NO"
result.df$diffexpressed[result.df$log2FoldChange > 1.5 & 
                        result.df$padj < 0.01] <- "UP"
result.df$diffexpressed[result.df$log2FoldChange < -1.5 & 
                        result.df$padj < 0.01] <- "DOWN"


ggplot(data = result.df, aes(x = log2FoldChange, y = -log10(pvalue), 
                             col = diffexpressed))+ 
  geom_point()+ theme_minimal()+
  geom_vline(xintercept = c(-1.5, 1.5), col = "black", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.01), col = "black", linetype = 'dashed') + 
  scale_color_manual(values = c("#00AFBB", "grey", "#FFDB6D"), 
  labels = c("Downregulated", "Not significant", "Upregulated"))

}

vplcano_plot()

```











